### DISCUSSION TITLE
Possiblity to directly use iterators ?

### DISCUSSION THREAD

**Discussion by magne4000** [remeda/remeda#967] [General]
I would want to do something like this:
```ts
const myMap = new Map<string, T>(...);
const iterator: IterableIterator<T> = myMap.values();

const first = R.pipe(
  iterator,
  R.first
);
```

Problem is `R.first` accepts a `T[]` and not an `IterableIterator<T>`, so my `first` var is of type `unknown` instead of being of type `T`;

**Comment by lstkz** [comment]
Only for pipe or for every function?

**Comment by magne4000** [comment]
Seems like almost every utils supporting iterable instead of arrays in [iter-tools](https://github.com/iter-tools/iter-tools/) could be candidate to this change.

**Comment by jaens** [comment]
This is such a bare minimum feature I am surprised it has not been implemented for this long... it basically makes this library unusable for working with ES `Map`s and `Set`s (which use `IterableIterator` for all their methods) since everything has to constantly be converted to `Array`s, it seems.

ES in general is also evolving towards using Iterator-like objects in new APIs.

Oh well, time to look for yet another alternative I guess...

**Comment by eranhirsch** [comment]
I agree this is a great addition to the library, I'm just not sure it's easy enough to be done without completely rewriting most of the library.

If anyone has any suggestion on how to go about this I'd be happy to discuss it.

**Comment by cjquines** [comment]
i think the main question is how the typing would work. `map` right now takes an array and returns an array. would we overload `map` with something that takes an iterable and returns an iterableiterator? that would technically be a breaking change, but it should be possible

i also don't think that converting everything to be iterator-based would be "completely rewriting most of the library". only 25 out of the 132 exported functions are lazy, and for many of them it wouldn't be too hard to convert. the hardest would of course be pipe, but if we converted everything to iterators, then the pipe implementation would also be much simpler (as we can runtime-inspect whether something is an iterator)

**Comment by cjquines** [comment]
i think it should be possible to write a pipe that operates on iterators:

```ts
R.iteratorPipe(
  iterator,
  R.map(x => x + 1),
  R.filter(x => x > 0),
);
```

the problem would be typing. either we let this return an array (which seems bad), or we want this to return an iterator

if we want this to return an iterator, then (someone check me on this) *every* operator in the pipe should be a `LazyOp` (or operate on iterators). idk how you check this on the type level. maybe you can ask everyone to explicitly use the `lazy` version:

```ts
R.iteratorPipe(
  iterator,
  R.map.lazy(x => x + 1),
  R.filter.lazy(x => x > 0),
);
```

this is worse ergonomics. (would it be against the remeda spirit? why else do we runtime purry between datafirst and datalast, rather than having separate function names?)

**Comment by eranhirsch** [comment]
It's unlikely I'll be able to invest the time in researching this as we are currently handling mainly feature parity to allow people to migrate off Ramda and Lodash, and clean up the API for a a v2 release.

My concern is that if the change requires rewriting a large chunk of the existing code then it would take a long time to make sure it's a valid solution.

**Comment by cjquines** [comment]
hm, please take a look at #520 – what do you think?

**Comment by eranhirsch** [comment]
Going over the pipe and purry functions lately, I have an idea for getting closer to using iterators. The main problem with the way we iterate lazily and how iterators work is that we "push" values into transformers instead of the transformer being able to pull them. What needs to be done:

* Add an "end-of-stream" symbol that could be pushed into a pipe. This would allow us to act not only when a value is pushed into the stream, which unlocks some other enhancements.
* Expand the "single" mode in pipes so that more of our functions can work lazily.
* Remove the "indexed" variants to simplify our codebase. Indexed affects how we lazily iterate because we maintain the index within the pipe.
* We need to bump our Typescript target to at least ES2015 so we can iterate over Iterables transparently without needing to manually handle them.

After all these changes, we can try to relax the APIs for some of our functions and see if it "just works". This would allow to put any iterable at the start of a pipe at least and have its values piped. The output might still be an object/array, not a map/set, but I think that's already a pretty good API.

**Comment by jaens** [comment]
_Side note:_ This sounds like a version of [Clojure's transducers](https://clojure.org/reference/transducers).

**Comment by somnicattus** [comment]
Hello.

I published remeda-compatible npm package rotery, which expands remeda with Iterators and AsyncIterators support.

https://github.com/somnicattus/rotery

**Comment by mb21** [comment]
API-wise, what do you think about a new module that exports all the functions with the same names, just for iterables and async iterables:

```ts
import * as R from "remeda"
import * as Ri from "remeda/iterable"
import * as Ra from "remeda/asyncIterable"

const iterable: Iterable<T> = new Map<string, T>(...).values();

// `first` would also get type `Iterable<T>`
const first = Ri.pipe(
  iterable,
  Ri.first
);
```

Note that some sync iterator helpers are standardized and [shipped in Chrome and Node.js](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/filter#browser_compatibility), but that [async iterator helpers](https://github.com/tc39/proposal-async-iterator-helpers) are held up. I don't think there are official TypeScript types for either of them yet.

Also, since there is only a prototype for Iterator and not Iterable, they unfortunately only work on Iterators. But that's not something Remeda would have to worry about, so supporting all iterables would be nicer I think.

**Comment by somnicattus** [comment]
Exporting new modules looks nice.

But supporting all `Iterable`s sounds a little too much.
Most people don't iterate on `Iterable`s such as `string`s or `ArrayBuffer`s. IMO, supporting only `IterableIterator`s is reasonable.
You always have a choice to get `IterableIterator`s from `Iterable`s by explicitly calling `iterable[Symbol.iterator]()`.

**Comment by somnicattus** [comment]
Iterator support potentially makes current lazy evaluation implementations obsolete, because iterators are always lazy.

**Comment by mb21** [comment]
@somnicattus thanks for the reply! Do you think accepting/emitting `Iterable`s is more difficult to implement than `IterableIterator`s? I assumed it's easier to emit, since it's a smaller interface, and accepting it shouldn't be a problem?

**Comment by somnicattus** [comment]
@mb21 
No, I don’t think it’s a problem of difficulty. `Iterable` and `IterableIterator` (and even `Iterator`) are essentially the same, except for their interfaces, and able to be transformed to each other.

I reconsidered and came to the conclusion that **accepting** `Iterable`s is not a bad idea. We can pass `IterableIterators` or `Iterables` as parameters of type `Iterable`. (So the best choice might be `Iterable | Iterator`s?)

However, when it comes to **emitting**, `IterableIterator`s is the obvious choice. Returned `IterableIterator`s can be treated as `Iterable`s, `Iterator`s, and of course `IterableIterator`s.

**Comment by jaens** [comment]
There's also the subtle question of side effects, since depending on the input/output type combination and internal "caching" details, input iterables could be consumed multiple times and output "side effects" could be executed multiple times (if the output iterable is consumed more than once).

**Comment by somnicattus** [comment]
> There's also the subtle question of side effects

Yes, `Iterator`s are essentially stateful objects and they will be "consumed". That's true, we need to think about it.

Even with the fact, there are some reasons why we want to support iterators and why we can do so.

- We sometimes prefer `Iterator`s over stateless objects because of "space complexity" (in other words, "memory usage").
- The results will be always the same with the "same state" inputs (it's referentially transparent, to some extent).
- The functionality does never affect the global state (`"sideEffect"` can be set to `false` and tree-shaking is allowed).
- We always have a choice to make the results stateless by doing `[ ...result ]`.
- There is already a `debounce` function in Remeda, which returns "stateful" results, but it has been already somehow accepted.
